File: Session 1/code/network-python/A2_thread.py

from threading import Thread
import time
import random
import display

def main():
    d = display.Display()

    time.sleep(0.5)
    n = 20
    for i in range(n):
        # start a producer every 300ms
        time.sleep(0.3)
        producer(d, i).start()
        # replace "start" by "run" above to not start a thread (and thus test as if you were not using threads...)


def producer(d, ind):
    def add_elements():
        for iteration in range(9):
            d.add_value(ind, 1)
            time.sleep(random.uniform(0, .5)) # wait randomly, up to 500ms
            d.render_clean()

    t = Thread(target=add_elements)
    return t


if __name__== "__main__":
  main()

# We have different result in each run, we just keep on adding more workers

File: Session 1/code/network-java/src/main/java/com/compnet/practical1/A2Thread.java

package com.compnet.practical1;

import java.util.logging.Level;
import java.util.logging.Logger;

/**
 *
 */
public class A2Thread {
    
    /**
     * @param args the command line arguments
     */
    public static void main(String[] args) {
        Display d = new Display();

        sleep(500, 0);
        int n = 20;
        for (int i = 0; i < n; i++) {
            // start a producer every 300ms
            sleep(300, 0);
            producer(d, i).start();
            // replace "start" by "run" above to not start a thread (and thus test as if you were not using threads...)
        }
    }

    private static Thread producer(final Display d, final int ind) {
        //* Java 8
        Runnable addElements = () -> {
            for (int i = 0; i < 9; i++) {
                d.add_value(ind, 1);
                sleep((int) (Math.random()*500), 0);
                d.render_clean();
            }
        };
        /*/// plain old java
        Runnable addElements = new Runnable() {
            @Override
            public void run() {
                for (int i = 0; i < 9; i++) {
                    d.add_value(ind, 1);
                    sleep((int) (Math.random()*500), 0);
                    d.render_clean();
                }
            }
        };
        //*/
        
        Thread t = new Thread(addElements);
        return t;
    }
    
    private static void sleep(long ms, int nanos) {
        try {
            Thread.sleep(ms, nanos);
        } catch (InterruptedException ex) {
            Logger.getLogger(Display.class.getName()).log(Level.SEVERE, null, ex);
        }
    }

}
